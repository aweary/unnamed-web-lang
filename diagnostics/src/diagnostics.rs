use source::diagnostics::{Diagnostic};
use source::filesystem::FileId;
use source::diagnostics::Span;

/// Both the lexer and parser return a Diagnostic for errors.
/// This might also be extended to other phases like typechecking
/// in the future.
pub type ParseResult<T> = std::result::Result<T, Diagnostic>;
