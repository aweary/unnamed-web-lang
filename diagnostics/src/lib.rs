#![warn(clippy::all)]

pub mod diagnostics;
pub mod error;
pub use diagnostics::*;
