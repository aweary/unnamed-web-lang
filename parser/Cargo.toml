[package]
name = "parser"
version = "0.1.0"
authors = ["Brandon Dail <brandondail@fb.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
graphql-parser = "0.2.3"
salsa = "0.14.1"
lexer = { path = "../lexer" }
syntax = { path = "../syntax" }
diagnostics = { path = "../diagnostics" }
source = { path = "../source" }
insta = { version = "0.16", features = ["glob"] }
log = "0.4.8"