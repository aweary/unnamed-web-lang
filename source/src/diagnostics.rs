use crate::filesystem::{FileId, FileSystem, Files};
pub use codespan::{ByteIndex, Span};
use codespan_reporting::diagnostic::{
    Diagnostic as CodespanDiagnostic, Label as CodespanLabel, LabelStyle, Severity,
};
pub use codespan_reporting::term::*;
use lsp_types::{
    Diagnostic as LspDiagnostic, DiagnosticSeverity, Position, Range as LspRange, Url,
};
use std::ops::Range;

// These diagnostic interfaces implement the same API as the codespan_reporting crate, except
// Label, which does not require a FileId when instantiated.

#[derive(Default)]
pub struct DiagnosticSet {
    pub diagnostics: Vec<Diagnostic>,
}

impl DiagnosticSet {
    pub fn add(&mut self, diagnostic: Diagnostic) {
        self.diagnostics.push(diagnostic)
    }
}

#[derive(Clone, Debug)]
pub struct Diagnostic {
    pub message: String,
    pub labels: Vec<Label>,
    severity: Severity,
}

impl Diagnostic {
    pub fn into_lsp(self, files: &FileSystem) -> (LspDiagnostic, Url) {
        // Find the first primary label
        let primary_label: &Label = self
            .labels
            .iter()
            .find(|label| label.style == LabelStyle::Primary)
            .expect("Diagnostics must have a primary label");
        let file = primary_label.file.expect("Labels must have files");
        // Get URI for the diagnostic from the file path
        let path = files.path_for_id(&file);
        let url = Url::from_file_path(path).unwrap();
        let range = {
            let range = &primary_label.range;
            // codespan-reporting Locations
            let start = files.location(file, range.start).unwrap();
            let end = files.location(file, range.end).unwrap();

            // lsp-types Positions
            let start = Position::new(
                (start.line_number - 1) as u64,
                (start.column_number - 1) as u64,
            );
            let end = Position::new((end.line_number - 1) as u64, (end.column_number - 1) as u64);
            LspRange::new(start, end)
        };
        let severity = match self.severity {
            Severity::Error | Severity::Bug => DiagnosticSeverity::Error,
            Severity::Warning => DiagnosticSeverity::Warning,
            Severity::Note => DiagnosticSeverity::Information,
            Severity::Help => DiagnosticSeverity::Hint,
        };
        let related_information = None;
        let tags = None;
        let code = None;
        let source: String = files.source(file).unwrap();
        let mut message = self.message.clone();
        message.push_str(": ");
        message.push_str(&primary_label.message);
        let diagnostic = LspDiagnostic::new(
            range,
            Some(severity),
            code,
            Some(source),
            message,
            related_information,
            tags,
        );
        (diagnostic, url)
    }

    pub fn emit_to_terminal(self, files: &FileSystem) {
        use termcolor::{ColorChoice, StandardStream};
        let writer = StandardStream::stderr(ColorChoice::Auto);
        let config = Config::default();
        // Convert to a codespan-reporting diagnostic
        let diagnostic = CodespanDiagnostic::new(self.severity)
            .with_message(self.message)
            .with_labels(
                self.labels
                    .into_iter()
                    .map(|label| {
                        let file = label.file.expect("Must have a file when reporting a label");
                        CodespanLabel::new(label.style, file, label.range)
                            .with_message(label.message)
                    })
                    .collect(),
            );
        emit(&mut writer.lock(), &config, &*files, &diagnostic).expect("Emitting");
    }

    pub fn error() -> Diagnostic {
        Diagnostic {
            message: String::new(),
            labels: vec![],
            severity: Severity::Error,
        }
    }

    pub fn with_message(mut self, message: impl Into<String>) -> Self {
        self.message = message.into();
        self
    }

    pub fn with_labels(mut self, labels: impl Into<Vec<Label>>) -> Self {
        self.labels = labels.into();
        self
    }

    pub fn for_file(mut self, file: FileId) -> Self {
        for mut label in &mut self.labels {
            if label.file.is_none() {
                label.file = Some(file);
            }
        }
        self
    }
}

#[derive(Clone, Debug)]
pub struct Label {
    pub file: Option<FileId>,
    pub style: LabelStyle,
    pub range: Range<usize>,
    pub message: String,
}

impl Label {
    pub fn primary(range: impl Into<Range<usize>>) -> Label {
        Label {
            file: None,
            style: LabelStyle::Primary,
            range: range.into(),
            message: String::new(),
        }
    }

    pub fn secondary(range: impl Into<Range<usize>>) -> Label {
        Label {
            file: None,
            style: LabelStyle::Secondary,
            range: range.into(),
            message: String::new(),
        }
    }

    pub fn with_message(mut self, message: impl Into<String>) -> Self {
        self.message = message.into();
        self
    }

    pub fn for_file(mut self, file: FileId) -> Self {
        self.file = Some(file);
        self
    }
}
