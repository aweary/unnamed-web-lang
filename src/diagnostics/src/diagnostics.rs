pub use codespan::*;
use codespan_reporting::diagnostic::Diagnostic;
pub use codespan_reporting::*;

/// Both the lexer and parser return a Diagnostic for errors.
/// This might also be extended to other phases like typechecking
/// in the future.
pub type ParseResult<T> = std::result::Result<T, Diagnostic>;
