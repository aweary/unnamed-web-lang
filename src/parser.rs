use ast::*;
use lexer::Lexer;
use token::*;

pub struct Parser {
    pub lexer: Lexer,
    // state: ParserState,
}

impl Parser {
    pub fn module(source: String) -> Self {
        let lexer = Lexer::new(source);
        // let state = ParserState::new();
        Parser { lexer }
    }

    fn peek(&mut self) -> Option<&Token> {
        self.lexer.peek_token()
    }

    pub fn skip_until<F>(&mut self, predicate: F)
    where
        F: Fn(Token) -> bool,
    {
        loop {
            match self.lexer.next_token() {
                Some(token) => {
                    if predicate(token) {
                        return;
                    }
                }
                None => return,
            }
        }
    }

    pub fn skip_assert(&mut self, token: Token) {
        assert_eq!(Some(token), self.lexer.next_token())
    }

    fn read_param_list(&mut self) -> Option<Vec<Param>> {
        match self.lexer.next_token() {
            // If the next token is an opening curly bracket, there
            // are no arguments and we assume this component takes no input
            Some(Token::LCurlyBracket) => None,
            // If the next token is an opening paren, there is
            // an argument list.
            Some(Token::LParen) => {
                let mut params: Vec<Param> = vec![];
                loop {
                    let name = self.read_ident();
                    // TODO this should be an unexpected token error
                    assert_eq!(Some(Token::Colon), self.lexer.next_token());
                    let kind = self.read_ident();
                    params.push(Param::new(name, kind));
                    match self.lexer.next_token() {
                        Some(Token::RParen) => {
                            break;
                            // End of the param list
                        }
                        Some(Token::Comma) => {
                            continue;
                            // There's another paramter
                        }
                        // Anything else is an unexpected token error
                        _ => unreachable!(),
                    }
                }
                Some(params)
            }
            _ => unreachable!("Unexpected token"),
        }
        // Next token should be the opening paren for the argument list
    }

    fn read_ident(&mut self) -> String {
        match self.lexer.next_token() {
            Some(Token::Ident(name)) => name,
            unknown @ _ => unreachable!("Expected identifier, found: {:?}", unknown),
        }
    }

    // Assumes that the current token is Reserved(Component)
    fn read_component_declaration(&mut self) -> Node {
        // Next token should be the identifier for the component definition
        let name = self.read_ident();
        // Read list of all params
        let params = self.read_param_list();
        // Everything from here until the closing bracket is part of the
        // component definition
        let body = self.read_component_body();
        let decl = ComponentDefinition::new(name, params, body);
        Node::ComponentDefinition(decl)
    }

    // A component is composed of a set of expressions/statements.
    fn read_component_body(&mut self) -> Vec<Expression> {
        let mut exprs = vec![];
        loop {
            match self.lexer.next_token() {
                // Statements starting with '<' are assumed to be
                // the beginning of JSX statements.
                Some(Token::LessThan) => {
                    let expr = Expression::JSXElement(self.read_jsx_element());
                    exprs.push(expr);
                    return exprs;
                }
                _ => {
                    continue;
                }
            }
        }
    }

    /**
     * read_expr
     * Reads any expresison, which is something that can evaluate to some value
     */
    fn read_expr(&mut self) -> Expression {
        self.skip_until(|token| token == Token::RCurlyBracket);
        Expression::NumericLiteral(42)
    }

    fn read_jsx_attributes(&mut self) -> Option<Vec<JSXAttribute>> {
        let mut attrs: Vec<JSXAttribute> = vec![];
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let attr = self.read_jsx_attribute();
                    attrs.push(attr);
                }
                Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                    break;
                }
                _ => {
                    unreachable!("Unknown token: {:?}", self.peek());
                }
            }
        }
        println!("read_jsx_attributes: {:?}", attrs);
        if attrs.is_empty() {
            None
        } else {
            Some(attrs)
        }
    }

    fn is_jsx_element_self_closing(&mut self) -> bool {
        match self.peek() {
            Some(&Token::GreaterThan) => {
                self.skip_assert(Token::GreaterThan);
                false
            }
            Some(&Token::ForwardSlash) => {
                self.skip_assert(Token::ForwardSlash);
                self.skip_assert(Token::GreaterThan);
                true
            }
            _ => {
                unreachable!();
            }
        }
    }

    fn read_jsx_element(&mut self) -> JSXElement {
        let tag_name = self.read_ident();
        let attributes = self.read_jsx_attributes();
        let self_closing = self.is_jsx_element_self_closing();
        let opening_element = JSXOpeningElement::new(tag_name.clone(), attributes, self_closing);
        println!("opening_element {:?}", opening_element);
        let (children, closing_element) = if self_closing {
            (None, None)
        } else {
            let children = self.read_jsx_children();
            let closing_element = self.read_jsx_closing_element();
            println!("closing_element {:?}", closing_element);
            println!("children {:?}", children);
            (Some(children), Some(closing_element))
        };
        let element = JSXElement::new(opening_element, closing_element, children);
        println!("element: {:#?}", element);
        element
    }

    fn read_jsx_closing_element(&mut self) -> JSXClosingElement {
        // Closing tag
        // self.skip_assert(Token::LessThan);
        // self.skip_assert(Token::ForwardSlash);
        let closing_tag_name = self.read_ident();
        // TODO parsing error
        // assert_eq!(tag_name, closing_tag_name);
        self.skip_assert(Token::GreaterThan);
        // TODO actually parse the closing tag
        JSXClosingElement::new(closing_tag_name)
    }

    fn read_jsx_text(&mut self) -> Box<JSXChildren> {
        let mut text = String::new();
        loop {
            match self.peek() {
                Some(&Token::Ident(_)) => {
                    let word = self.read_ident();
                    // TODO respect whitespace when lexing instead of this hack
                    text = text + " " + &word;
                }
                Some(&Token::LessThan) | Some(&Token::LCurlyBracket) => {
                    break;
                }
                _ => {
                    unreachable!();
                }
            }
        }
        println!("read_jsx_text: {:?}", text);
        Box::new(JSXChildren::JSXText(text))
    }

    fn read_jsx_children(&mut self) -> Vec<Box<JSXChildren>> {
        let mut children: Vec<Box<JSXChildren>> = vec![];
        loop {
            match self.peek() {
                // JSXText
                Some(&Token::Ident(_)) => {
                    children.push(self.read_jsx_text());
                },
                Some(&Token::LessThan) => {
                    // A '<' token could be the start of another element
                    // or the closing element for the enclosing element
                    self.skip_assert(Token::LessThan);
                    match self.peek() {
                        Some(&Token::Ident(_)) => {
                            // An identifier means the start of a new element
                            let element = self.read_jsx_element();
                            children.push(Box::new(JSXChildren::JSXElement(element)));
                        }
                        Some(&Token::ForwardSlash) => {
                            self.skip_assert(Token::ForwardSlash);
                            return children;
                            // A '/' indicates a closing element
                        }
                        _ => {
                            unreachable!();
                        }
                    }
                }
                _ => break,
            }
        }
        children
        // let child = match token {
        //     // TODO JSX parsers respect the whitespace inside tags. Need to output tokens
        //     Some(Token::Ident(word)) => {
        //         JSXChildren::JSXText(word)
        //     }
        //     Some(Token::LessThan) => {
        //         let element = self.read_jsx_element();
        //         JSXChildren::JSXElement(element)
        //     }
        //     // TODO handle JSXExpressionContainer
        //     token @ _ => {
        //         unreachable!("Unexpected token: {:?}", token);
        //     }
        // };
        // Box::new(child)
    }

    /**
     * Reads the expression part of a JSX attribute.
     */
    fn read_jsx_attribute_value(&mut self) -> JSXAttributeValue {
        // An explicit assignment. The assigned value can
        // either be a string literal or an identifier in curly brackets
        match self.lexer.next_token() {
            Some(Token::String(literal)) => JSXAttributeValue::StringLiteral(literal),
            Some(Token::LCurlyBracket) => {
                let expr = self.read_expr();
                JSXAttributeValue::Expression(expr)
            }
            token @ _ => unreachable!("Unknown token: {:?}", token),
        }
    }

    /**
     * read_jsx_attr
     *
     * A JSX Attribute comes in two forms:
     *
     * 1. An explicit attribute where the value is passed
     *    in inline (<div id="1" />)
     * 2. An implicit attribute where we assume the attribute
     *    name is a reference to some identifier
     *    (<div id>)
     */
    fn read_jsx_attribute(&mut self) -> JSXAttribute {
        let name = self.read_ident();

        let foo = match self.peek() {
            Some(&Token::Assign) => {
                self.skip_assert(Token::Assign);
                let attribute_value = self.read_jsx_attribute_value();
                JSXAttribute::new(name, attribute_value)
            }
            Some(&Token::Ident(_)) | Some(&Token::ForwardSlash) | Some(&Token::GreaterThan) => {
                let attr_ident = Identifier::new(name.clone());
                let attr_ident_expr = Expression::Identifer(attr_ident);
                let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
                JSXAttribute::new(name, attr_value)
            }
            _ => {
                unreachable!();
            }
        };
        foo
        // println!("read_jsx_attribute:name {:?}", name);
        // match self.lexer.next_token() {
        //     // Explicit assignment
        //     Some(Token::Assign) => {
        //         let attr_value = self.read_jsx_attribute_value();
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     // Shorthand form
        //     Some(Token::Ident(_)) | Some(Token::ForwardSlash) | Some(Token::GreaterThan) => {
        //         println!("shorthand form, creating the identifier and stuff");
        //         let attr_ident = Identifier::new(name.clone());
        //         let attr_ident_expr = Expression::Identifer(attr_ident);
        //         let attr_value = JSXAttributeValue::Expression(attr_ident_expr);
        //         JSXAttribute::new(name, attr_value)
        //     }
        //     token @ _ => unreachable!("Unknown character: {:?}", token),
        // }
    }

    fn read_type_def(&mut self) -> Node {
        // Next token is the identifier for the type definition.
        let name = self.read_ident();
        let def = TypeDefinition::new(name);
        // Just read everything until we reach the end for now...
        loop {
            if self.lexer.next_token() == Some(Token::RCurlyBracket) {
                break;
            }
        }
        Node::TypeDefinition(def)
    }

    /**
     * Inside a module, top level nodes are currently restricted to component
     * and function definitions. This means you cannot have other expressions
     * that are executed when the module is evalulated. All components and
     * functions must be lazily evaluated.
     */
    pub fn next_node(&mut self) -> Node {
        let token = self.lexer.next_token().unwrap();
        match token {
            Token::Reserved(ReservedWord::Component) => self.read_component_declaration(),
            Token::Reserved(ReservedWord::Type) => self.read_type_def(),
            _ => Node::Text,
        }
    }
}
