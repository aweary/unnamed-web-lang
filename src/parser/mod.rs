mod decl;
mod expr;
mod stmt;
mod context;

use decl::DeclParser;
// use expr::ExprParser;
// use stmt::StmtParser;
pub use context::{ParsingContext, ExprContext};


use crate::error::ParseError;
use crate::lexer::Lexer;
use crate::result::Result;
use crate::ast::{Module, Program};
use crate::symbol::Symbol;
use crate::token::{Token, TokenKind};
use std::rc::Rc;
use std::cell::RefCell;

// temporary use so that the Rust compiler checks this module
// use crate::ir::*;

pub fn parse_program<'a>(source: &'a str, mut ctx: &'a mut ParsingContext) -> Result<Program> {
    let mut parser = Parser::new(&source, &mut ctx);
    parser.parse()
}

pub struct Parser<'a> {
    lexer: Lexer<'a>,
    ctx: &'a mut ParsingContext,
}

impl<'a> Parser<'a> {
    pub fn new(source: &'a str, ctx: &'a mut ParsingContext) -> Self {
        let lexer = Lexer::new(&source);
        Parser { lexer, ctx }
    }

    fn next_token(&mut self) -> Result<Token> {
        match self.lexer.next_token(&mut self.ctx) {
            Ok(token) => {
                Ok(token)
            }
            Err(_) => Err(ParseError::LexError),
        }
    }

    fn peek_token(&mut self) -> Result<&Token> {
        match self.lexer.peek_token(&mut self.ctx) {
            Ok(token) => Ok(token),
            Err(_) => Err(ParseError::LexError),
        }
    }

    fn eat(&mut self, kind: TokenKind) -> Result<bool> {
        if self.peek_token()?.kind == kind {
            self.expect(kind)?;
            Ok(true)
        } else {
            Ok(false)
        }
    }

    fn expect(&mut self, kind: TokenKind) -> Result<Token> {
        let token = self.next_token()?;
        if token.kind != kind {
            Err(ParseError::UnexpectedToken(token))
        } else {
            Ok(token)
        }
    }

    fn ident(&mut self) -> Result<Symbol> {
        let token = self.next_token()?;
        match token.kind {
            TokenKind::Ident(sym) => Ok(sym),
            _ => Err(ParseError::UnexpectedToken(token)),
        }
    }

    /**
     * This is currently an alias to `ident`, but a type
     * identifier has different syntax. For example, polymorphic
     * types have type parameters.
     */
    fn type_ident(&mut self) -> Result<Symbol> {
        self.ident()
    }

    pub fn parse(&mut self) -> Result<Program> {
        // top-level statements
        let stmts = self.decl_list()?;
        let module = Module { stmts };
        let modules = vec![module];
        let program = Program { modules };
        Ok(program)
    }
}
