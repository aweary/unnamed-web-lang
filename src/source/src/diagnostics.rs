pub use codespan_reporting::diagnostic::{Diagnostic, LabelStyle, Label};
pub use codespan_reporting::term::*;
pub use crate::index::{ByteIndex};
pub use crate::span::Span;


