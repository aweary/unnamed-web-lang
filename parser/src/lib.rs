#![warn(clippy::all)]

mod parser;

pub mod test_utils;
pub use crate::parser::*;

#[cfg(test)]
mod tests {
    use crate::parser::Parser;
    use diagnostics::ParseResult as Result;
    use insta::assert_json_snapshot;
    use std::fs;
    use std::path::PathBuf;
    use syntax::{ast, precedence::Precedence};

    fn parse_stmts(source: &str) -> Vec<ast::Stmt> {
        let mut parser = Parser::new(source);
        parser.stmt_list().unwrap()
    }

    fn parse_module(source: &str) -> ast::Module {
        let mut parser = Parser::new(source);
        parser.parse_module().unwrap()
    }

    fn parse_expr(source: &str) -> ast::Expr {
        let mut parser = Parser::new(source);
        parser.expr(Precedence::NONE).unwrap()
    }

    fn expr_snapshot(name: &str, input: &str) {
        let name = format!("expr_{}", name);
        assert_json_snapshot!(name, parse_expr(input))
    }

    fn stmts_snapshot(name: &str, input: &str) {
        let name = format!("stmt_{}", name);
        assert_json_snapshot!(name, parse_stmts(input))
    }

    #[test]
    fn literal_snapshots() {
        expr_snapshot("integer", "1234");
        expr_snapshot("float", "12.34");
        expr_snapshot("long_float", "23.32342123423");
        expr_snapshot("string", "\"Hello\"");
        expr_snapshot("boolean_true", "true");
        expr_snapshot("boolean_false", "false");
        expr_snapshot("array", "[1, 2, 3]");
        expr_snapshot("object", "{a: 1, b: 2}");
    }

    #[test]
    fn let_stmt_snapshots() {
        stmts_snapshot(
            "numbers",
            "
          let a = 1
          let b = 2
          let c = 3.4
        ",
        );

        stmts_snapshot(
            "ident_names",
            "
        let a = 1
        let longerName = 2
        let longerNameWith_underscore = 3
        let _ = 4
        let _leadingUnderscore = 5
        ",
        )
    }

    #[test]
    fn fixtures() {
        let input = include_str!("./inputs/function_simple.dan");
        assert_json_snapshot!("function_simple_fixture", parse_module(input));
    }

}
