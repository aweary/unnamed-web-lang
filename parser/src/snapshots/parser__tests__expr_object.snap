---
source: parser/src/lib.rs
expression: parse_expr(input)
---
{
  "id": 0,
  "kind": {
    "Object": [
      [
        {
          "name": {
            "unstable_source": "a"
          },
          "span": {
            "start": 1,
            "end": 2
          }
        },
        {
          "id": 0,
          "kind": {
            "Lit": {
              "span": {
                "start": 4,
                "end": 5
              },
              "kind": {
                "Number": {
                  "unstable_source": "1"
                }
              }
            }
          },
          "span": {
            "start": 4,
            "end": 5
          },
          "ty": null
        }
      ],
      [
        {
          "name": {
            "unstable_source": "b"
          },
          "span": {
            "start": 7,
            "end": 8
          }
        },
        {
          "id": 0,
          "kind": {
            "Lit": {
              "span": {
                "start": 10,
                "end": 11
              },
              "kind": {
                "Number": {
                  "unstable_source": "2"
                }
              }
            }
          },
          "span": {
            "start": 10,
            "end": 11
          },
          "ty": null
        }
      ]
    ]
  },
  "span": {
    "start": 0,
    "end": 12
  },
  "ty": null
}
